### PR TITLE
focus: rework and untangle existing focus management logic in the sdk

### PR DESCRIPTION
**Pull request by mimecuvalo**
Focus management is really scattered across the codebase. There's sort of a battle between different code paths to make the focus the correct desired state. It seemed to grow like a knot and once I started pulling on one thread to see if it was still needed you could see underneath that it was accounting for another thing underneath that perhaps wasn't needed.

The impetus for this PR came but especially during the text label rework, now that it's much more easy to jump around from textfield to textfield. It became apparent that we were playing whack-a-mole trying to preserve the right focus conditions (especially on iOS, ugh). 

This tries to remove as many hacks as possible, and bring together in place the focus logic (and in the darkness, bind them).

## Places affected
- [x] `useEditableText`: was able to remove a bunch of the focus logic here. In addition, it doesn't look like we need to save the selection range anymore.
  - lingering footgun that needed to be fixed anyway: if there are two labels in the same shape, because we were just checking `editingShapeId === id`, the two text labels would have just fought each other for control
- [x] `useFocusEvents`: nixed and refactored — we listen to the store in `FocusManager` and then take care of autoFocus there
- [x] `useSafariFocusOutFix`: nixed. not necessary anymore because we're not trying to refocus when blurring in `useEditableText`. original PR for reference: https://github.com/tldraw/brivate/pull/79
- [x] `defaultSideEffects`: moved logic to `FocusManager`
- [x] `PointingShape` focus for `startTranslating`, decided to leave this alone actually.
- [x] `TldrawUIButton`: it doesn't look like this focus bug fix is needed anymore, original PR for reference: https://github.com/tldraw/tldraw/pull/2630
- [x] `useDocumentEvents`: left alone its manual focus after the Escape key is hit
- [x] `FrameHeading`: double focus/select doesn't seem necessary anymore
- [x] `useCanvasEvents`: `onPointerDown` focus logic never happened b/c in `Editor.ts` we `clearedMenus` on pointer down
  - [x] `onTouchStart`: looks like `document.body.click()` is not necessary anymore

## Future Changes
- [ ] a11y: work on having an accessebility focus ring
- [ ] Page visibility API: (https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) events when tab is back in focus vs. background, different kind of focus
- [ ] Reexamine places we manually dispatch `pointer_down` events to see if they're necessary.
- [ ] Minor: get rid of `useContainer` maybe? Is it really necessary to have this hook? you can just do `useEditor` → `editor.getContainer()`, feels superfluous.

## Methodology
Looked for places where we do:
- `body.click()`
- places we do `container.focus()`
- places we do `container.blur()`
- places we do `editor.updateInstanceState({ isFocused })`
- places we do `autofocus`
- searched for `document.activeElement`

### Change Type

- [x] `sdk` — Changes the tldraw SDK
- [ ] `dotcom` — Changes the tldraw.com web app
- [ ] `docs` — Changes to the documentation, examples, or templates.
- [ ] `vs code` — Changes to the vscode plugin
- [ ] `internal` — Does not affect user-facing stuff

- [ ] `bugfix` — Bug fix
- [ ] `feature` — New feature
- [x] `improvement` — Improving existing features
- [ ] `chore` — Updating dependencies, other boring stuff
- [ ] `galaxy brain` — Architectural changes
- [ ] `tests` — Changes to any test code
- [ ] `tools` — Changes to infrastructure, CI, internal scripts, debugging tools, etc.
- [ ] `dunno` — I don't know

### Test Plan

- [x] run test-focus.spec.ts
- [x] check MultipleExample
- [x] check EditorFocusExample
- [x] check autoFocus
- [x] check style panel usage and focus events in general
- [x] check text editing focus, lots of different devices, mobile/desktop

### Release Notes

- Focus: rework and untangle existing focus management logic in the SDK

### CODE REVIEW

File: apps/dotcom/src/components/CursorChatBubble.tsx
Comment (mimecuvalo) [inline apps/dotcom/src/components/CursorChatBubble.tsx:128]: this double raf doesn't seem necessary... but maybe i'm missing something.

File: apps/dotcom/src/components/PeopleMenu/UserPresenceEditor.tsx
Comment (mimecuvalo) [inline apps/dotcom/src/components/PeopleMenu/UserPresenceEditor.tsx:55]: drive-by rename to make `autoFocus` across the SDK consistent

File: packages/editor/src/lib/hooks/useCanvasEvents.ts
Comment (mimecuvalo) [inline packages/editor/src/lib/hooks/useCanvasEvents.ts:43]: this logic path had bitrotted, as mentioned in the PR description

File: packages/editor/src/lib/hooks/useCanvasEvents.ts
Comment (mimecuvalo) [inline packages/editor/src/lib/hooks/useCanvasEvents.ts:100]: doesn't look like this is necessary anymore...

File: packages/tldraw/src/lib/defaultSideEffects.ts
Comment (mimecuvalo) [inline packages/tldraw/src/lib/defaultSideEffects.ts:5]: moved to FocusManager

File: packages/tldraw/src/lib/shapes/frame/components/FrameHeading.tsx
Comment (mimecuvalo) [inline packages/tldraw/src/lib/shapes/frame/components/FrameHeading.tsx:62]: this double raf doesn't seem necessary anymore...

File: packages/tldraw/src/lib/shapes/shared/useEditableText.ts
Comment (mimecuvalo) [inline packages/tldraw/src/lib/shapes/shared/useEditableText.ts:89]: we were doing a lot of hoops and somersaults to account for what `TldrawUiButtonPicker` should have been managing

File: packages/tldraw/src/lib/ui/components/primitives/TldrawUiButtonPicker.tsx
Comment (mimecuvalo) [inline packages/tldraw/src/lib/ui/components/primitives/TldrawUiButtonPicker.tsx:58]: The logic for what useEditableText was trying to accomplish was moved here

File: packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx
Comment (mimecuvalo) [inline packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx:17]: As mentioned in the PR desc., it looks like this isn't needed anymore...

**Comment by mimecuvalo** [comment]
@MitjaBezensek @SomeHats actually hold off on reviewing - there's something i have to fix first

**Comment by mimecuvalo** [comment]
@MitjaBezensek @SomeHats ok! all good now

File: packages/tldraw/src/test/Editor.test.tsx
Comment (mimecuvalo) [inline packages/tldraw/src/test/Editor.test.tsx:449]: hm, i dunno why this test needed to be paused actually. i stared at it but not seeing the difference. @SomeHats if you have more historical context here maybe lemme know.

Comment (SomeHats) [reply packages/tldraw/src/test/Editor.test.tsx:449]: So in the past, if the tldraw editor was focused, and then focus left (e.g. went to some input somewhere else on the page) we would update the instance state appropriately. I honestly have no idea where we were/are doing this though, but the test still seems to make some conceptual sense to me?

**Comment by SomeHats** [comment]
this is amazing work <3 would love to see tests for some of the places where we've removed workarounds if they don't already exist but also yikes this stuff is hard to test

File: apps/examples/e2e/tests/test-focus.spec.ts
Comment (MitjaBezensek) [inline apps/examples/e2e/tests/test-focus.spec.ts:142]: I'm not sure if understand why the draw tool should become unselected here?

File: packages/tldraw/src/test/Editor.test.tsx
Comment (MitjaBezensek) [reply packages/tldraw/src/test/Editor.test.tsx:449]: Searching for `focusout` it seems like we only used it in `useSafariFocusOutFix`, which has been completely removed in this PR. Not sure we even handle this event anywhere, so the test might just be obsolete?

Comment (mimecuvalo) [reply packages/tldraw/src/test/Editor.test.tsx:449]: yeah, but the weird thing is, this test was testing not just Safari 🤔

File: packages/editor/src/lib/editor/managers/FocusManager.ts
Comment (mimecuvalo) [reply packages/editor/src/lib/editor/managers/FocusManager.ts:19]: ah, that's a good point - lemme double check. prbly better to not have a raf in there, just in case.

**Comment by mimecuvalo** [comment]
> this is amazing work <3 would love to see tests for some of the places where we've removed workarounds if they don't already exist but also yikes this stuff is hard to test

good point + thx! and yeah, hah, hard to test - but yeah, worth it - lemme try to add some more 👍

File: apps/examples/e2e/tests/test-focus.spec.ts
Comment (mimecuvalo) [reply apps/examples/e2e/tests/test-focus.spec.ts:142]: yeah, it took me a little bit to figure out - it's a quirk of this test.
it's because the `autoFocus` prop is changing, so in `TldrawEditorWithReadyStore`→ `useLayoutEffect`, the changing prop causes the editor to be reset. you wouldn't _normally_ change the `autoFocus` prop to change focus, you'd want to programatically change the editor you already have.

Comment (mimecuvalo) [reply apps/examples/e2e/tests/test-focus.spec.ts:142]: that said, maybe i should update `MultipleExample` to not change props but actually update the instance state to more properly address this.

Comment (MitjaBezensek) [reply apps/examples/e2e/tests/test-focus.spec.ts:142]: > yeah, it took me a little bit to figure out - it's a quirk of this test. it's because the `autoFocus` prop is changing, so in `TldrawEditorWithReadyStore`→ `useLayoutEffect`, the changing prop causes the editor to be reset. you wouldn't _normally_ change the `autoFocus` prop to change focus, you'd want to programatically change the editor you already have.

ouch 😂 At least for me it definitely sounds like a property you set once and should not change, so the change in behaviour is fine.

Yeah, updating the example sounds good.

File: packages/tldraw/src/test/Editor.test.tsx
Comment (SomeHats) [reply packages/tldraw/src/test/Editor.test.tsx:449]: is this not generally desired behaviour though? if focus leaves the editor in the DOM, the internal state of the editor should reflect that, right?

File: apps/examples/e2e/tests/test-focus.spec.ts
Comment (mimecuvalo) [reply apps/examples/e2e/tests/test-focus.spec.ts]: i reworked MultipleExample to not changethe prop

Comment (SomeHats) [reply apps/examples/e2e/tests/test-focus.spec.ts]: i don't feel suuuuuper strongly here - happy to leave as is - but in general, I would look to the browser/react for how attributes mirroring built-in ones should behave.

In both of those, i believe setting this attribute on a pre-existing element does nothing, but it definitely doesn't clear the state of the items. in the browser, i think it's respected on page load, and in react it's respected when an element is first mounted.

File: packages/tldraw/src/test/Editor.test.tsx
Comment (mimecuvalo) [reply packages/tldraw/src/test/Editor.test.tsx:449]: Well, to be clear though, this is leaving focus from a child of the editor, not the editor directly. I think if it leaves focus from a child it would just go to a parent element possibly? It's sort of depends on where the person clicks next or tabs to next.

File: apps/examples/e2e/tests/test-focus.spec.ts
Comment (mimecuvalo) [reply apps/examples/e2e/tests/test-focus.spec.ts]: This is a good point, I reworked this little bit so that TldrawEditor only takes the initial auto focus value. Changing it shouldn't affect the editor.

**Comment by mimecuvalo** [comment]
added a couple more tests (for edit→edit case and for the style picker+maintaining focus in textarea case). lemme know if there's another case to add but i think we should be good to go!

File: apps/dotcom/src/components/CursorChatBubble.tsx
Comment (ds300) [reply apps/dotcom/src/components/CursorChatBubble.tsx:128]: yeah your version works in chrome, safari and firefox. cc @TodePond do you remember if this double raf was intentional?

File: packages/editor/src/lib/editor/Editor.ts
Comment (ds300) [inline packages/editor/src/lib/editor/Editor.ts:8053]: I wonder whether this should set the `instanceState.isFocused` to true if not already?

Comment (ds300) [reply packages/editor/src/lib/editor/Editor.ts:8053]: or really maybe it should be the job of the focus manager

File: packages/editor/src/lib/hooks/useCanvasEvents.ts
Comment (ds300) [reply packages/editor/src/lib/hooks/useCanvasEvents.ts:101]: did you test this on iOS? seems fine on android

File: packages/tldraw/src/lib/shapes/shared/useEditableText.ts
Comment (ds300) [inline packages/tldraw/src/lib/shapes/shared/useEditableText.ts:25]: you sure this is still working on iOS?

File: packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx
Comment (ds300) [reply packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx:17]: are you sure? the document.activeElement still reverts to the `body` when the button becomes inactive in my tests. admittedly that no longer seems to prevent keyboard shortcuts from working but maybe it could cause other issues?

**Review by ds300 — APPROVED** [review]
I don't feel super qualified to review this but things all seem to work on mac and android in my testing so 👍🏼 

The only thing I'm feeling pause on is the behaviour/api of the focus manager.

- should calling focusManager.focus update the instanceState.isFocused property, and same for the .blur method?
- should we keep the manager itself private and just have a .focus and .blur method on the editor that call the underlying manager methods? I know how steve feels about namespaces.

File: packages/editor/src/lib/hooks/useCanvasEvents.ts
Comment (mimecuvalo) [reply packages/editor/src/lib/hooks/useCanvasEvents.ts:101]: yes, i've been testing on the iOS simulator. will definitely test again once it's on staging.

File: packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx
Comment (mimecuvalo) [reply packages/tldraw/src/lib/ui/components/primitives/Button/TldrawUiButton.tsx:17]: i think i'm specifically making sure that the original PR (https://github.com/tldraw/tldraw/pull/2630) is working. if it causes other issues, then i think i would rather that we know what those issues are vs. this covering them up potentially (as a fix for something unrelated)

File: packages/editor/src/lib/editor/Editor.ts
Comment (mimecuvalo) [reply packages/editor/src/lib/editor/Editor.ts:8053]: i think possibly: that it's tricky. if there are multiple editors on the page, i could see an argument that you might want to temporarily translate/move a shape on one editor (Editor A), but the focus remains on a separate editor (Editor B). and that just because you focus Editor A, it doesn't mean that you want Editor A to take over control from Editor B. but perhaps that's overthinking it?

File: packages/tldraw/src/lib/shapes/shared/useEditableText.ts
Comment (mimecuvalo) [reply packages/tldraw/src/lib/shapes/shared/useEditableText.ts:25]: i've been testing on the iOS sim - will definitely test on staging once it lands to double check!

**Comment by mimecuvalo** [comment]
> should calling focusManager.focus update the instanceState.isFocused property, and same for the .blur method?

yeah, it's a good question - like i said above: i think possibly: that it's tricky. if there are multiple editors on the page, i could see an argument that you might want to temporarily translate/move a shape on one editor (Editor A), but the focus remains on a separate editor (Editor B). and that just because you focus Editor A, it doesn't mean that you want Editor A to take over control from Editor B. but perhaps that's overthinking it?

> should we keep the manager itself private and just have a .focus and .blur method on the editor that call the underlying manager methods? I know how steve feels about namespaces.

fair point - i think i can make the focus manager be private actually. (to be clear, no one is calling `blur()` explicitly at the moment, it's only used in FocusManager at the moment)

**Pull request merged**
